### PR TITLE
feat(api): 1523 add market prices to emp state and call to list known market prices

### PR DESCRIPTION
**Pull request by daywiss**
Signed-off-by: David Adams <david@umaproject.org>

**Motivation**

https://app.clubhouse.io/uma-project/story/1523/connect-aggregator-api-to-our-api-or-sdk-for-pulling-in-the-best-price-the-token-is-trading-across-dexs

**Summary**

Adds infrastructure to fetch latest market price from 0x api, and keep it relatively up to date. This then gets added to emp state when queried as `tokenMarketPrice`. 

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

https://app.clubhouse.io/uma-project/story/1523/connect-aggregator-api-to-our-api-or-sdk-for-pulling-in-the-best-price-the-token-is-trading-across-dexs

### CODE REVIEW

File: packages/api/src/apps/api/README.md [hunk 1]
```diff
@@ -17,6 +17,9 @@ cryptowatchApiKey=
 tradermadeApiKey=
 quandlApiKey=
 defipulseApiKey=
+
+# zrx price feed gets us market prices on tokens
```
Comment (chrismaree) [inline packages/api/src/apps/api/README.md:21]:
```suggestion
# 0x price feed gets us market prices on tokens
```

File: packages/api/src/libs/queries.ts [hunk 2]
```diff
@@ -51,6 +54,13 @@ export default (appState: Dependencies) => {
     // [timestamp, price], returns just price
     return synthPrices.latest[empAddress][1];
   }
+  async function getLatestMarketPrice(address: string, currency: "usdc" = "usdc") {
```
Comment (chrismaree) [inline packages/api/src/libs/queries.ts:57]: how do decimals work with this exactly? is it always returned in the decimals of the `currency`?

File: packages/api/src/services/market-prices.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+type Dependencies = Pick<AppState, "zrx" | "marketPrices" | "collateralAddresses" | "syntheticAddresses">;
+
+// market prices are pulled from the 0x matcha api
+export default function (config: Config, appState: Dependencies) {
+  // these prices will be quoted against usdc by default, but can be specified as address or symbol
+  const { zrx, marketPrices, collateralAddresses, syntheticAddresses } = appState;
+  // this is hardcoded for now since it differs from the standard currency symbol usd
+  const currency = "usdc";
+
+  // does not do any queries, just a helper to mutate the latest price table
```
Comment (chrismaree) [inline packages/api/src/services/market-prices.ts:16]: I might be missing something that came in a previous PR, but where does the fetching actually happen?

**Review by chrismaree — APPROVED** [review]

**Review by nicholaspai — APPROVED** [review]
Optional nit: rename zrx to 0x which is the name of the protocol, instead of the name of the protocol token

File: packages/api/src/libs/queries.ts (hunk 2 above)
Comment (daywiss) [reply packages/api/src/libs/queries.ts:57]: prices are returned as a float and on return i convert it to we

File: packages/api/src/services/market-prices.ts (hunk 3 above)
Comment (daywiss) [reply packages/api/src/services/market-prices.ts:16]: yea, there was a previous pr to this https://github.com/UMAprotocol/protocol/pull/3203.  This uses the 0x api

**Comment by daywiss** [comment]
> Optional nit: rename zrx to 0x which is the name of the protocol, instead of the name of the protocol token

The name of the project is annoying becuase js variables can't start with a number, which is why i ended up using zrx. Open to alternatives though @nicholaspai

File: packages/api/src/libs/queries.ts (hunk 2 above)
Comment (chrismaree) [reply packages/api/src/libs/queries.ts:57]: right buy what are the decimals? i.e lets say the price is 1000 USDC per eth. is it represented as "1000" or is it "1000e6"  using the decimals for USDC?

**Review by chrismaree — APPROVED** [review]

**Pull request merged**